src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/TagHelperBinder.cs

```diff
@@ -14,30 +14,80 @@ namespace Microsoft.AspNetCore.Razor.Language;
 /// </summary>
 internal sealed class TagHelperBinder
 {
-    private readonly Dictionary<string, HashSet<TagHelperDescriptor>> _registrations;
+    private readonly ImmutableArray<TagHelperDescriptor> _catchAllDescriptors;
+    private readonly Dictionary<string, ImmutableArray<TagHelperDescriptor>> _tagNameToDescriptorsMap;
```

Comment:
Should this be some kind of immutable dictionary? It looks like we don't change it after initialization.

Reply:
Right. We don't change it after initialization, so it's safe to use for lookup. `Dictionary<,>` is much faster for lookup and takes up less spece than `ImmutableDictionary<,>`. And, `FrozenDictionary<,>` is slower to create, so it is usually reserved for static data. Since this is instance data that is created during compiler phases, I decided that it'd be better to aim for faster to create/populate and faster for lookup.

Reply:
Makes sense, thanks, consider leaving a comment.

Reply:
Will do.